osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/SendDestinationAndMethodConverter.java

```diff
@@ -0,0 +1,101 @@
+      new EnumMap<>(MessageTypeDto.class);
+
+  static {
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.TCP, 0);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.UDP, 1);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.FTP, 2);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.SMTP, 3);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.SMS, 4);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.HDLC, 5);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.M_BUS, 6);
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.ZIG_BEE, 7);
+    /*
+     * Could be 200..255, use first value as long as no more information is
+     * available.
+     */
+    ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.put(TransportServiceTypeDto.MANUFACTURER_SPECIFIC, 200);
+
+    ENUM_VALUE_PER_MESSAGE_TYPE.put(MessageTypeDto.A_XDR_ENCODED_X_DLMS_APDU, 0);
+    ENUM_VALUE_PER_MESSAGE_TYPE.put(MessageTypeDto.XML_ENCODED_X_DLMS_APDU, 1);
+    /*
+     * Could be 128..255, use first value as long as no more information is
+     * available.
+     */
+    ENUM_VALUE_PER_MESSAGE_TYPE.put(MessageTypeDto.MANUFACTURER_SPECIFIC, 128);
+  }
+
+  @Override
+  public DataObject convert(
+      final SendDestinationAndMethodDto source,
+      final Type<? extends DataObject> destinationType,
+      final MappingContext context) {
+    if (source == null) {
+      return null;
+    }
+
+    final List<DataObject> sendDestinationAndMethodElements = new ArrayList<>();
+
+    // add service
+    final int enumValueTransportServiceType =
+        this.getEnumValueTransportServiceType(source.getTransportService());
+    sendDestinationAndMethodElements.add(
+        DataObject.newEnumerateData(enumValueTransportServiceType));
+
+    // add destination
+    sendDestinationAndMethodElements.add(
+        DataObject.newOctetStringData(source.getDestination().getBytes(StandardCharsets.US_ASCII)));
+
+    // add message
+    final int enumValueMessageType = this.getEnumValueMessageType(source.getMessage());
+    sendDestinationAndMethodElements.add(DataObject.newEnumerateData(enumValueMessageType));
+
+    return DataObject.newStructureData(sendDestinationAndMethodElements);
+  }
+
+  private int getEnumValueTransportServiceType(final TransportServiceTypeDto transportServiceType) {
+    final Integer enumValue = ENUM_VALUE_PER_TRANSPORT_SERVICE_TYPE.get(transportServiceType);
+    if (enumValue == null) {
+      throw new AssertionError("Unknown TransportServiceType: " + transportServiceType);
+    }
+    return enumValue;
+  }
+
+  private int getEnumValueMessageType(final MessageTypeDto messageType) {
+    final Integer enumValue = ENUM_VALUE_PER_MESSAGE_TYPE.get(messageType);
+    if (enumValue == null) {
+      throw new AssertionError("Unknown MessageType: " + messageType);
+    }
+    return enumValue;
+  }
```

Comment:
I'm wondering if these methods will still be needed after adding the dlmsEnumValue to the DTO classes?
Or should we keep these methods?
What would be the correct place for keeping and maintaining the dlms enum values? In the DTOs moving between domain and protocol adapter or just in the protocol adapter?

Reply:
No, they are not needed anymore. Since there are already more DTOs that containg the enum values for DLMS (e.g. SignalQualityDto), I think the dto is a good place for the values.